libraries/AP_HAL/Device.h

```diff
@@ -78,6 +78,12 @@ class AP_HAL::Device {
         }
     }
 
+    /*
+     * Change device address. Note that this is the 7 bit address, it
+     * does not include the bit for read/write. Only works on I2C
+     */
+    virtual void set_address(uint8_t address) {};
```

Comment:
it doesn't really make sense to change a SPI address, does it? this should remain on the I2CDevice afaics

Reply:
it is needed so we can have a generic driver that works with both bus types, but still allows set_address() on i2c.

Reply:
you have the bus type and you can act accordingly on where it is needed.

Reply:
on some drivers what I do is to have 2 constructors, one receiving I2CDevice and the other receiving SPIDevice. Then there's the common method that both call. Chances are high setting the address is not the only change you need to do anyway.  And in cases you don't want to differentiate the constructor you can still resort to
```c++
if (dev->bus_type == AP_HAL::Device::BUS_TYPE_I2C) { ... }
```